### DISCUSSION TITLE
How to combine multiple ONNX models sequentially?

### DISCUSSION THREAD

**Discussion by ghost** [onnx/onnx#5096] [Q&A]
Hello

May I know how do we combine multiple ONNX models?
These are the 3 models that I need to merge at the moment
1) Model 1 -> sklearn OneHotEncoder
2) Model 2 -> Tensorflow Keras Model
3) Model 3 -> [sklearn Isotonic Regression](https://scikit-learn.org/stable/modules/generated/sklearn.isotonic.IsotonicRegression.html) (Since sklearn-onnx doesn't have a registered converter for this, I have created a custom converter with the underlying logic used

Here the output of Model 1 will be the input of Model 2 and output of Model 2 will be the input of Model 2

I tried to use onnx.compose.merge_models but ended up getting the error - 
`ValueError: Can't merge two models with different operator set ids for a given domain`

**Comment by cbourjau** [comment]
I would recommend doing it using the new [Spox](https://github.com/Quantco/spox) library which has excellent [`inline` support](https://spox.readthedocs.io/en/latest/guides/inline.html#Converter-libraries). It should also support updating the individual models to a common opset version. (Full disclosure: I'm involved in that project)

**Comment by ghost** [comment]
Thank you @cbourjau for your suggestion. I will try the Spox library

I just have one more question, at the moment skl2onnx doesn't support [Isotonic Regression](https://scikit-learn.org/stable/modules/generated/sklearn.isotonic.IsotonicRegression.html) and developing a [new converter](https://onnx.ai/sklearn-onnx/auto_tutorial/plot_icustom_converter.html) has been a bit challenging. Do you have any recommendations/suggestions for it?

**Reply by cbourjau** [reply]
"Converters" are essentially lazy re-implementations of the `predict` and/or `transform` logic. You will have to express the logic using ONNX operators. Looking at the source of the `IsotonicRegression` class it seems straightforward until the call to `interp1d` where things turn a bit complex...

Let's take a step back:
The tricky thing with converters is also how to organize your logic around them. Our (internal) converters are always written as classes mirroring the sklearn ones. You can take a look at the docs [here](https://spox.readthedocs.io/en/latest/guides/converter.html) on how to organize your converters.

For the problem at hand, I recommend a test-driven approach: Copy the body of the `IsotonicRegression` class into your new converter class (e.g. `SpoxIsotonicRegression`) and set up a test calling `build` as described in the docs. Then attempt the conversion which will obviously fail because you have not used any onnx operators yet. Work your way through your `SpoxIsotonicRegression` replacing numpy operations with operations from `spox.opset.ai.onnx.v18` (as described in the docs). 

At the end you will have a converter class that composes well with your first two models using Spox's `inline` functionality. 

I have written converters in both Spox and skl2onnx. As stated earlier I am biased to the latter, but I am quite confident in saying it will be way easier to write this converter in Spox than it is in skl2onnx. You can reuse existing skl2onnx converters whenever available through `inline`.

I hope this helps!
